swag/bank.py

```diff
@@ -301,6 +308,226 @@ def set_timezone(self, user, timezone):
+
+    def get_active_cagnotte(self, cagnotte_idx) -> Cagnotte:
+        if cagnotte_idx >= 0 and cagnotte_idx < self.swagdb.cagnotte_number():
+            cagnotte = self.swagdb.get_cagnotte_from_index(cagnotte_idx)
+            if cagnotte.cagnotte_activation == True:
+                return cagnotte
+
+        raise NoCagnotteRegistered(cagnotte_idx)
+
+    def get_all_active_cagnotte(self):
+        return [
+            cagnotte
+            for cagnotte in self.swagdb.get_cagnotte_infos()
+            if cagnotte.cagnotte_activation == True
+        ]
+
+    def get_cagnotte_history(self, cagnotte_idx):
+        return [
+            transaction
+            for transaction in self.swagdb.blockchain
+            if concerns_cagnotte(cagnotte_idx, transaction)
+        ]
+
+    def payer_a_cagnotte(
+        self, donator_account_discord_id: int, cagnotte_idx: int, montant
+    ):
+        cagnotte = self.get_active_cagnotte(cagnotte_idx)
+        donator_account = self.swagdb.get_account(donator_account_discord_id)
+        # On regarde le type de la cagnotte pour pouvoir correctement choisir les fonctions qui devront être utiliser
+        if cagnotte.get_info().cagnotte_currency == "$wag":
+
+            # Check if the value of $wag is correct regardless its propriety
+            if not isinstance(montant, int) or montant < 0:
+                raise InvalidSwagValue
+
+            # Check if the donator have enough $wag:
+            if (
+                self.get_account_info(donator_account_discord_id).swag_balance - montant
+                < 0
+            ):
+                raise NotEnoughSwagInBalance(donator_account_discord_id)
+
+            # Making the donation
+            donator_account.swag_balance -= montant
+            cagnotte.cagnotte_balance += montant
+
+        elif cagnotte.get_info().cagnotte_currency == "$tyle":
+
+            # Check if the value of $tyle is correct regardless its propriety
+            if not isinstance(montant, (int, float, Decimal)) or montant < 0:
+                raise InvalidStyleValue
+
+            # Check if the donator have enough $tyle:
+            if (
+                self.get_account_info(donator_account_discord_id).style_balance
+                - montant
+                < 0
+            ):
+                raise NotEnoughStyleInBalance
+
+            donator_account.style_balance -= montant
+            cagnotte.cagnotte_balance += montant
+
+        # Write donation in the blockchain
+        self.swagdb.blockchain.append(
+            (
+                utcnow().datetime,
+                TransactionType.DONATION,
+                (
+                    donator_account.get_info().id,
+                    cagnotte.get_info().cagnotte_id,
+                    montant,
+                    cagnotte.get_info().cagnotte_currency,
+                ),
+            )
+        )
+
+        if donator_account_discord_id not in cagnotte.get_info().cagnotte_participant:
```

Comment:
Resolved in : https://github.com/BobbySoftwares/Bybbot/pull/7/commits/676277344a85e73ee79eff0731dea5e07dff9faf

I'm stupid, I never use set and I shouldn't, because their propriety can be very usefull